base/abstractarray.jl

```diff
@@ -1522,14 +1522,33 @@ _isdisjoint(as::Tuple, bs::Tuple) = !(as[1] in bs) && _isdisjoint(tail(as), bs)
 """
     Base.dataids(A::AbstractArray)
 
-Return a tuple of `UInt`s that represent the mutable data segments of an array.
+Return a tuple of `UInt`s that identify the mutable data segments of an array.
 
-Custom arrays that would like to opt-in to aliasing detection of their component
-parts can specialize this method to return the concatenation of the `dataids` of
-their component parts.  A typical definition for an array that wraps a parent is
-`Base.dataids(C::CustomArray) = dataids(C.parent)`.
+These values are used to determine if two arrays might share memory with [`Base.mightalias`](@ref).
+The default implementation recursively combines the `dataids` of all fields of the struct.
+
+Custom arrays only need to implement a custom `dataids` method if:
+
+* they wish to ignore some fields (with non-empty `dataids`) in aliasing considerations;
+    for example this can be the case if an array is used to store intentionally-shared
+    metadata or other data that is not mutated by `setindex!`
+
+* or they depend upon non-array fields (with empty `dataids`) to define their indexable
+    contents that they wish to include in aliasing considerations.
 """
-dataids(A::AbstractArray) = (UInt(objectid(A)),)
+function dataids(A::AbstractArray)
+    @inline
+    if @generated
+        :(ids = tuple($([:(dataids(getfield(A, $i))...) for i in 1:fieldcount(A)]...)))
+    else
+        ids = _splatmap(dataids, ntuple(i -> getfield(A, i), Val(nfields(A))))
```

Comment:
Do we need to check all registered packages for undef fields in `AbstractArray` before merging this?

Reply:
The PkgEval runs didn't come up with any such arrays to my reading, which is indeed a bit surprising (but welcome)!

Reply:
Haha maybe its literally only `Tridiagonal`. 

How certain are we than package tests actually call `dataids` somewhere on all `AbstractArray`?

Reply:
Good question — they'll hit it if they touch any of the builtin implementations of broadcasting, `view`, or `reshape`... which I'd rate as fairly likely (even if not explicitly as a targeted test, they'd probably do at least one of those in some implementation).  I'd think, anyhow.

Reply:
Yes it sounds unlikely none of those are called.